src/downloader/downloader.cpp

```diff
@@ -5,34 +5,195 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#include <fcntl.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+
 #include "downloader.hpp"
+#include "log.hpp"
+
+Downloader::Downloader()
+    : mDownloadRequester(nullptr)
+    , mThread()
+    , mMutex()
+    , mWaitDownload(mMutex)
+    , mErrProcessImageRequest()
+    , mRequestedPath()
+    , mDownloadResults()
+    , mTimer(Downloader::DownloadTimeoutCallback, this)
+    , mRequestID(0)
+{
+}
+
+Downloader::~Downloader()
+{
+    mWaitDownload.NotifyOne();
+}
+
+void Downloader::DownloadTimeoutCallback(void* self)
+{
+    auto* downloader = static_cast<Downloader*>(self);
+
+    LOG_ERR() << "Download timeout";
+
+    downloader->setErrorAndNotify(aos::ErrorEnum::eTimeout);
+}
 
 aos::Error Downloader::Init(DownloadRequesterItf& downloadRequester)
 {
+    LOG_DBG() << "Initialize CM client";
+
     mDownloadRequester = &downloadRequester;
 
     return aos::ErrorEnum::eNone;
 }
 
 aos::Error Downloader::Download(const aos::String& url, const aos::String& path, aos::DownloadContent contentType)
 {
-    (void)url;
-    (void)path;
-    (void)contentType;
+    LOG_DBG() << "Download: " << url;
 
-    return aos::ErrorEnum::eNone;
+    aos::UniqueLock lock(mMutex);
+    mDownloadResults.Clear();
+
+    auto err = mTimer.Create(Downloader::cDownloadTimeout);
+    if (!err.IsNone()) {
+        LOG_DBG() << "Create timer failed " << strerror(errno);
+        return AOS_ERROR_WRAP(err);
+    }
+
+    mRequestedPath = path;
+    mErrProcessImageRequest = aos::ErrorEnum::eNone;
+
+    err = mThread.Run([this, &url, contentType](void*) {
+        mErrProcessImageRequest = mDownloadRequester->SendImageContentRequest({url, ++mRequestID, contentType});
+    });
+    if (!err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    mThread.Join();
+
+    if (!mErrProcessImageRequest.IsNone()) {
+        return mErrProcessImageRequest;
+    }
+
+    mWaitDownload.Wait();
```

Comment:
This is error-prone way. You should always have wait condition otherwise dead lock is possible.